### PR TITLE
feat: [DHIS2-13299][DHIS2-16291] Related Stages

### PR DESCRIPTION
**Pull request by eirikhaugstulen**
TECH-summary:
This widget checks if there are a relationship type of program stage to program stage. If there are on, there is a widget appearing underneath the New Enrollment event form.

How to test:
1. Create a new relationship type
2. This type should be `Event in program or program stage` => `Event in program or program stage` in the same program
3.  Make sure to check the sharing setting so that the user can capture data

image

**This PR is made up of several issues who have all been reviewed. Please only check the latest commit**

### CODE REVIEW

**Review by JoakimSM — APPROVED** [review]

**Comment by simonadomnisoru** [comment]
Tested successfully on 2.41, 2.40.4, 2.39.6, 2.38.7 versions

File: src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.epics.js [hunk 1]
```diff
@@ -26,3 +28,24 @@ export const updateEventFailedEpic = (action$: InputObservable) =>
             return batchActions([saveFailed(), rollbackEnrollmentEvent(eventId)]);
         }),
     );
+export const saveEventAndCompleteEnrollmentSucceededEpic = (action$: InputObservable) =>
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.epics.js:31]: Why were these 2 epics moved from `EnrollmentAddEventPage.epics` into `EnrollmentEditEventPage.epics.js`? 🤔 
I wasn't able to trigger them, what are the steps to do so? Also the `eventId` is known when editing so there is no need to send it in `commitEnrollmentAndEvents(meta.uid, eventId)`

File: src/core_modules/capture-core/components/WidgetEventEdit/EditEventDataEntry/editEventDataEntry.actions.js [hunk 2]
```diff
@@ -24,6 +24,8 @@ export const actionTypes = {
     EVENT_SCHEDULE_ERROR: 'ScheduleEvent.UpdateScheduleEventError',
     START_CREATE_NEW_AFTER_COMPLETING: 'WidgetEventEdit.StartCreateNewAfterCompleting',
     EVENT_SAVE_ENROLLMENT_COMPLETE_REQUEST: 'WidgetEventEdit.EventSaveAndEnrollmentCompleteRequest',
+    EVENT_SAVE_ENROLLMENT_COMPLETE_SUCCESS: 'NewEnrollmentEventPage.EventSaveAndEnrollmentCompleteSuccess',
```
[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/WidgetEventEdit/EditEventDataEntry/editEventDataEntry.actions.js:27]: I think the reason for this was that originally I thought that it should just be one save and one dispatch for each result, i.e. one save, one commit and one rollback, that handles all scenarios. However, I could not delete it as editEvent was still using it, and therefore the refactor. This is also the reason for your third comment. Do you want me to remove it?

File: src/core_modules/capture-core/components/Pages/EnrollmentAddEvent/EnrollmentAddEventPageDefault/EnrollmentAddEventPageDefault.container.js [hunk 3]
```diff
@@ -62,20 +65,28 @@ export const EnrollmentAddEventPageDefault = ({
...
-            history.push(`enrollment?${buildUrlQueryString({ programId, orgUnitId, teiId, enrollmentId })}`);
-        },
-        [dispatch, history, programId, orgUnitId, teiId, enrollmentId],
-    );
+        ({ enrollments, events, linkMode }) => {
+            if (linkMode && linkMode === RelatedStageModes.ENTER_DATA) return;
+
+            const nowClient = fromClientDate(new Date());
+            const nowServer = new Date(nowClient.getServerZonedISOString());
+            const updatedAt = moment(nowServer).format('YYYY-MM-DDTHH:mm:ss');
```
[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Pages/EnrollmentAddEvent/EnrollmentAddEventPageDefault/EnrollmentAddEventPageDefault.container.js:73]: Could probably be removed from getAddEventEnrollmentServerData, but not sure about CompleteModal? Is that not used in other places as well, like the edit page?

File: src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.epics.js (hunk 1 above)
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.epics.js:31]: No, see my original comment further up. Do you think it safe to remove?

Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/Pages/EnrollmentEditEvent/EnrollmentEditEventPage.epics.js:31]: Yes, I bealive is safe to remove these 2 epics (they are not being triggered at the moment). When editing the event the commit and rollback are currently being handled in `editEventDataEntry.epics` by [saveEditedEventSucceededEpic](https://github.com/dhis2/capture-app/blob/master/src/core_modules/capture-core/components/WidgetEventEdit/EditEventDataEntry/editEventDataEntry.epics.js#L171) and [saveEditedEventFailedEpic](https://github.com/dhis2/capture-app/blob/master/src/core_modules/capture-core/components/WidgetEventEdit/EditEventDataEntry/editEventDataEntry.epics.js#L202)

File: src/core_modules/capture-core/components/WidgetEventEdit/EditEventDataEntry/editEventDataEntry.actions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/WidgetEventEdit/EditEventDataEntry/editEventDataEntry.actions.js:27]: They can be removed together with the 2 epics from `EnrollmentEditEventPage.epics.js`

File: src/core_modules/capture-core/components/Pages/EnrollmentAddEvent/EnrollmentAddEventPageDefault/EnrollmentAddEventPageDefault.container.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/Pages/EnrollmentAddEvent/EnrollmentAddEventPageDefault/EnrollmentAddEventPageDefault.container.js:73]: Good point, `CompleteModal` is used by 2 different pages. Probably `updatedAt` should also be calculated only once for all edited events. I don't know what the most sensible approach is, but calculating it multiple times throughout the code depending on how the event is added/edited doesn't seem right 🤔

[the remark this answers is not in the document]
Comment (JoakimSM) [reply src/core_modules/capture-core/components/Pages/EnrollmentAddEvent/EnrollmentAddEventPageDefault/EnrollmentAddEventPageDefault.container.js:73]: Makes sense to calculate once, but if it is hard to determine/hard to refactor, let's make a separate ticket.

[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/Pages/EnrollmentAddEvent/EnrollmentAddEventPageDefault/EnrollmentAddEventPageDefault.container.js:73]: https://dhis2.atlassian.net/browse/DHIS2-17100

**Review by JoakimSM — APPROVED** [review]

**Review by simonadomnisoru — APPROVED** [review]

**Pull request merged**